client/src/components/requestPanel/RequestPanel.tsx

```diff
@@ -20,34 +20,74 @@ import KVEditor from '../kvEditor';
 import UriBar from '../uriBar';
 import styles from './RequestPanel.module.css';
 
-const defaultParam = {
-  key: '',
-  value: '',
-};
+function getParamsFromUri(uri: string, params?: Array<KVRow>): Array<KVRow> {
+  const paramString = uri.split('?')[1];
+  if (!paramString) {
+    return params?.filter((param) => param.isEnabled === false) ?? [];
+  }
+
+  const uriParams = paramString.split('&').map((kv) => {
+    const [k, ...v] = kv.split('='); // ...v with v.join('=') handle cases where the value contains '='
+    return {
+      key: k,
+      value: v.join('='),
+    };
+  });
+
+  if (!params) {
+    return uriParams;
+  }
 
-function shouldAppendNewRow(params: Array<KVRow>): boolean {
-  if (params.length === 0) return true;
-  const { key, value } = params[params.length - 1];
-  return key !== '' || value !== '';
+  const newParams: KVRow[] = [];
+
+  let indexEnabledParams = 0;
+  for (const [_, param] of params.entries()) {
+    if (param.isEnabled === false) {
+      newParams.push(param);
+    } else {
+      const uriParam = uriParams[indexEnabledParams];
+      if (!uriParam) {
+        console.error('params and URI params out of sync');
+        newParams.push({ key: '', value: '' });
+      } else {
+        newParams.push(uriParam);
+      }
+      indexEnabledParams++;
+    }
+    if (uriParams[indexEnabledParams]) {
+      console.error('params and URI params out of sync');
+    }
+    // add remaining URI params to newParams in case they go out of sync
+    for (let i = indexEnabledParams; i < uriParams.length; i++) {
```

Comment:
we don't need this one.

Reply:
Why do you think that? I think we need this indicator to find and debug out-of-sync situations where uriParams >  params.
We can make it a warning if you want.